noxfile.py

```diff
@@ -59,3 +64,151 @@ def docs(session):
+# Helpers
+# -----------------------------------------------------------------------------
+def _get_version_from_arguments(arguments):
+    """Checks the arguments passed to `nox -s release`.
+
+    Only 1 argument that looks like a version? Return the argument.
+    Otherwise, raise a ValueError describing what's wrong.
+    """
+    if len(arguments) != 1:
+        raise ValueError("Expected exactly 1 argument")
+
+    version = arguments[0]
+    parts = version.split(".")
+
+    if len(parts) != 2:
+        # Not of the form: YY.N
+        raise ValueError("not of the form: YY.N")
+
+    if not all(part.isdigit() for part in parts):
+        # Not all segments are integers.
+        raise ValueError("non-integer segments")
+
+    # All is good.
+    return version
+
+
+def _check_working_directory_state(session):
+    """Check state of the working directory, prior to making the release.
+    """
+    should_not_exist = ["build/", "dist/"]
+
+    bad_existing_paths = list(filter(os.path.exists, should_not_exist))
+    if bad_existing_paths:
+        session.error(f"Remove {', '.join(bad_existing_paths)} and try again")
+
+
+def _check_git_state(session, version_tag):
+    """Check state of the git repository, prior to making the release.
+    """
+    # Ensure the upstream remote pushes to correct URL
+    allowed_upstreams = [
+        "git@github.com:pypa/packaging.git",
+        "https://github.com/pypa/packaging.git",
+    ]
+    result = subprocess.run(
+        ["git", "remote", "get-url", "--push", "upstream"],
+        capture_output=True,
+        encoding="utf-8",
+    )
+    if result.stdout.rstrip() not in allowed_upstreams:
+        session.error(f"git remote `upstream` is not one of {allowed_upstreams}")
+    # Ensure we're on master branch for cutting a release.
+    result = subprocess.run(
+        ["git", "rev-parse", "--abbrev-ref", "HEAD"],
+        capture_output=True,
+        encoding="utf-8",
+    )
+    if result.stdout != "master\n":
+        session.error(f"Not on master branch: {result.stdout!r}")
+
+    # Ensure there are no uncommitted changes.
+    result = subprocess.run(
+        ["git", "status", "--porcelain"], capture_output=True, encoding="utf-8"
+    )
+    if result.stdout:
+        print(result.stdout, end="")
+        session.error(f"The working tree has uncommitted changes")
+
+    # Ensure this tag doesn't exist already.
+    result = subprocess.run(
+        ["git", "rev-parse", version_tag], capture_output=True, encoding="utf-8"
+    )
+    if not result.returncode:
+        session.error(f"Tag already exists! {version_tag} -- {result.stdout!r}")
+
+    # Back up the current git reference, in a tag that's easy to clean up.
+    _release_backup_tag = "auto/release-start-" + str(int(time.time()))
+    session.run("git", "tag", _release_backup_tag, external=True)
+
+
+def _bump(session, *, version, file, kind):
+    session.log(f"Bump version to {version!r}")
```

Comment:
Why only logging here?

Reply:
I think this is comparing to `_check_working_directory_state` and `_check_git_state`.

That's because this call is actually making changes to the sources / git repository. I'd guessed that it'd be a good idea to have clear indicators of what changes are being made as we go (in the output itself).